extension/solarwindsextension/internal/heartbeat.go

```diff
@@ -0,0 +1,138 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/extension"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+	"sync"
+	"time"
+)
+
+type Heartbeat struct {
+	logger *zap.Logger
+
+	cancel           context.CancelFunc
+	startShutdownMtx sync.Mutex
+
+	metric        *UptimeMetric
+	exporter      *Exporter
+	collectorName string
+}
+
+var alreadyRunningError = errors.New("heartbeat already running")
+
+func NewHeartbeat(ctx context.Context, set extension.Settings, cfg *Config) (*Heartbeat, error) {
+	set.Logger.Debug("Creating Heartbeat")
+	h := &Heartbeat{
+		logger:        set.Logger,
+		metric:        newUptimeMetric(set.Logger),
+		collectorName: cfg.CollectorName,
+	}
+
+	var err error
+	h.exporter, err = newExporter(ctx, set, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return h, nil
+}
+
+func (h *Heartbeat) Start(ctx context.Context, host component.Host) error {
+	h.startShutdownMtx.Lock()
+	defer h.startShutdownMtx.Unlock()
+
+	h.logger.Debug("Starting Heartbeat routine")
+	if h.cancel != nil {
+		return alreadyRunningError
+	}
+
+	err := h.exporter.start(ctx, host)
+	if err != nil {
+		return err
+	}
+
+	var loopCtx context.Context
+	loopCtx, h.cancel = context.WithCancel(context.Background())
+	go h.loop(loopCtx)
+	return nil
+}
+
+func (h *Heartbeat) Shutdown(ctx context.Context) error {
+	h.startShutdownMtx.Lock()
+	defer h.startShutdownMtx.Unlock()
+
+	h.logger.Debug("Stopping Heartbeat routine")
+	if h.cancel == nil {
+		// already stopped
+		return nil
+	}
+	h.cancel()
+	h.cancel = nil
+	return h.exporter.shutdown(ctx)
+}
+
+func (h *Heartbeat) loop(ctx context.Context) {
+	tick := time.NewTicker(30 * time.Second)
+	defer tick.Stop()
+
+	// Start beat
+	if err := h.generate(ctx); err != nil {
+		h.logger.Error("Generating heartbeat failed", zap.Error(err))
+	}
+
+	for {
+		select {
+		case <-tick.C:
+			if err := h.generate(ctx); err != nil {
+				h.logger.Error("Generating heartbeat failed", zap.Error(err))
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+
+}
+
+func (h *Heartbeat) generate(ctx context.Context) error {
+	h.logger.Debug("Generating heartbeat")
+	md := pmetric.NewMetrics()
+
+	if err := h.metric.add(ctx, md); err != nil {
+		return err
+	}
+
+	for i, rms := 0, md.ResourceMetrics(); i < rms.Len(); i++ {
+		rm := rms.At(i)
+		if err := h.decorateResourceAttributes(rm.Resource()); err != nil {
+			return err
+		}
+	}
+
+	return h.exporter.push(ctx, md)
+}
+
+func (h *Heartbeat) decorateResourceAttributes(resource pcommon.Resource) error {
+	if h.collectorName != "" {
+		resource.Attributes().PutStr("collector_name", h.collectorName)
```

Comment:
Looking at the [Resource](https://opentelemetry.io/docs/specs/semconv/resource/)/[Attributes](https://opentelemetry.io/docs/specs/semconv/general/attribute-naming/) naming conventions, we might have more specific round of discussion about the newly added functionality naming. For starters, it might be recommended to go with namespace approach (`collector.name`, ...) but there is more to it, metric naming included.

Reply:
For the record, as discussed offline, let's name:
- metric as `sw.otelcol.uptime`
- res. attribute as `sw.otelcol.collector.name`
To avoid any potential future clashing with pretty general `collector` namespace, since these changes are bound to sw context.

Reply:
Renamed in 5c987888becb7af855d10c766bd913e6cac1ecd1